### PR TITLE
Allow sortResults to work for 2D and 3D

### PR DESCRIPTION
**Pull request by masterleinad**
This should fix the issue mentioned in https://github.com/arborx/ArborX/pull/189#discussion_r347570498.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by aprokop** [comment]
Is there nothing in Kokkos to allow permute/swap elements?

**Comment by masterleinad** [comment]
> Is there nothing in Kokkos to allow permute/swap elements?

These are all implementation details of `Kokkos::BinSort` I basically copied here.

**Comment by masterleinad** [comment]
I am happy to move this function into a public namespace like `ArborX::Utils` or the like if that is a good compromise for everyone.

**Comment by dalg24** [comment]
> External application (DTK) should not impose any requirements on our `Details` namespace.

> move this function into a public namespace like ArborX::Utils or the like

Having these living in the `Details::` namespace is our stating that they are not part of the ArborX API and we make no promises about not breaking your code if you use it.   Renaming the namespace won't change a thing about it.

I would like to point out a number of things:
* ArborX uses stuff like `Kokkos::Impl::SpaceAccessibility` and has been using more in the past
* DTK pulls much more from ArborX than this sorting utility.  Remember it provides a number of STL-like algorithms that operates on Views, etc.
* DTK uses implementation details from Tpetra and from Intrepid2 as well

It is what it is and it is and we are not going to fix all in this PR (if we ever need to fix it).

**Comment by dalg24** [comment]
I am not convinced about a068912 
Do you have a direct need to use `sortResults()` outside the distributed search tree?

Also `sortMultipleViews()` is misleading.  It is not what it does.  It permutes all the trailing views in the order that sort the first view.  Note that I am not advocating that `sortResults()` is a great name either but it lives next to `countResults()` and `filterResults()` and it is only needed there as far as I know.

File: test/tstDetailsSortUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(sort_results_2d, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  std::array<int, 5> ids{4, 2, 1, 3, 5};
+  Kokkos::View<int *, DeviceType> ids_device("ids", 5);
+  auto ids_host = Kokkos::create_mirror_view(ids_device);
+  for (int i = 0; i < 5; ++i)
+    ids_host[i] = ids[i];
+  Kokkos::deep_copy(ids_device, ids_host);
```
Comment (dalg24) [inline test/tstDetailsSortUtils.cpp]: We know how to do this more efficiently now
```C++
Kokkos::deep_copy(ids_device, Kokkos::View<int*, Kokkos::HostSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>(ids.data(), ids.size()));
```

File: src/details/ArborX_DetailsSortUtils.hpp [hunk 2]
```diff
@@ -114,6 +114,106 @@ Kokkos::View<size_t *, Kokkos::Device<Kokkos::Cuda, MemorySpace>> sortObjects(
...
+  static void copy(DstViewType const &dst, size_t i_dst, SrcViewType const &src,
+                   size_t i_src)
+  {
+    for (int j = 0; j < (int)dst.extent(1); j++)
+      dst(i_dst, j) = src(i_src, j);
+  }
+};
+
+template <class DstViewType, class SrcViewType>
+struct CopyOp<DstViewType, SrcViewType, 3>
```
Comment (dalg24) [inline src/details/ArborX_DetailsSortUtils.hpp]: Please confirm that we use this in DTK.

Comment (masterleinad) [reply src/details/ArborX_DetailsSortUtils.hpp]: No, we only use 1D and 2D views. I am restoring the same functionality we had before (which is the same as in `Kokkos::BinSort`).

Comment (masterleinad) [reply src/details/ArborX_DetailsSortUtils.hpp]: No, we only use 1D and 2D views. I am restoring the same functionality we had before (which is the same as in Kokkos::BinSort).

**Comment by masterleinad** [comment]
> I am not convinced about [a068912](https://github.com/arborx/ArborX/commit/a068912b87cf3b0cc4739176a22625d6ea17d1b6)
> Do you have a direct need to use `sortResults()` outside the distributed search tree?

No, but it is also not that specific to the distributed search tree implementation. `sortObjects` could also just have been defined in `details/ArborX_DetailsBatchedQueries.hpp` before using it in `sortResults`. The main motivation for a068912 was to address https://github.com/arborx/ArborX/pull/189#discussion_r343167952. I don't feel strong about that. If you and @aprokop come to the agreement that you don't want this, I am happy to drop the commit.

**Comment by dalg24** [comment]
```C++
View v1{1,2,3}; // assume that is valid syntax to initialize a rank-1 view from an initializer list
View v2{3,2,1};
View v3{2,1,3};
sortMultipleViews(v1, v2, v3); // effect?
// (A) v1 == {1,2,3} && v2 == {1,2,3} && v3 == {1,2,3}
// (B) v1 == {1,2,3} && v2 == {2,1,3} && v3 == {3,2,1}
// (C) v1 == {1,2,3} && v2 == {3,2,1} && v3 == {2,1,3}
// (D) ...
```

**Comment by dalg24** [comment]
The variadic function to apply the permutation to all trailing views (more than just one) is of little use outside the distributed search tree.

That said reusing the sorting utility and adding the function to apply the permutation to a single view to a common header is not a bad idea.

File: src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp [hunk 3]
```diff
@@ -502,7 +476,8 @@ void DistributedSearchTreeImpl<DeviceType>::sortResults(
   Kokkos::deep_copy(keys_clone, keys);
   auto const permutation = ArborX::Details::sortObjects(keys_clone);
 
-  applyPermutations(permutation, other_views...);
+  auto dummy = {
+      (ArborX::Details::applyPermutations(permutation, other_views), 0)...};
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp]: This is being too clever.  I am even surprised the compiler does not complain about forming a reference to void in the initializer list.
At the very least you need to comment.

File: src/details/ArborX_DetailsSortUtils.hpp [hunk 4]
```diff
@@ -114,6 +114,67 @@ Kokkos::View<size_t *, Kokkos::Device<Kokkos::Cuda, MemorySpace>> sortObjects(
...
+  {
+    for (int j = 0; j < dst.extent(1); j++)
+      for (int k = 0; k < dst.extent(2); k++)
+        dst(i_dst, j, k) = src(i_src, j, k);
+  }
+};
+} // namespace PermuteHelper
+
+template <typename PermutationView, typename View>
+void applyPermutations(PermutationView const &permutation, View &view)
```
Comment (dalg24) [inline src/details/ArborX_DetailsSortUtils.hpp]: If the variadic version does not exist, should it be `applyPermutation()` singular?

File: src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp (hunk 3 above)
Comment (Rombur) [reply src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp]: This line throws a warning in clang:
```
In file included from /var/jenkins/workspace/ArborX_PR-197/test/tstDetailsDistributedSearchTreeImpl.cpp:14:
/var/jenkins/workspace/ArborX_PR-197/src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp:479:8: warning: unused variable 'dummy' [-Wunused-variable]
  auto dummy = {
       ^
```

File: src/details/ArborX_DetailsSortUtils.hpp [hunk 5]
```diff
@@ -114,6 +114,67 @@ Kokkos::View<size_t *, Kokkos::Device<Kokkos::Cuda, MemorySpace>> sortObjects(
...
+
+template <class DstViewType, class SrcViewType>
+struct CopyOp<DstViewType, SrcViewType, 3>
+{
+  KOKKOS_INLINE_FUNCTION
+  static void copy(DstViewType const &dst, size_t i_dst, SrcViewType const &src,
+                   size_t i_src)
+  {
+    for (int j = 0; j < dst.extent(1); j++)
+      for (int k = 0; k < dst.extent(2); k++)
```
Comment (Rombur) [inline src/details/ArborX_DetailsSortUtils.hpp]:
```
/var/jenkins/workspace/ArborX_PR-197/test/tstDetailsDistributedSearchTreeImpl.cpp:119:449:   required from here
/var/jenkins/workspace/ArborX_PR-197/src/details/ArborX_DetailsSortUtils.hpp:154:19: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (int j = 0; j < dst.extent(1); j++)
/var/jenkins/workspace/ArborX_PR-197/src/details/ArborX_DetailsSortUtils.hpp:155:19: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       for (int k = 0; k < dst.extent(2); k++)
```

File: src/details/ArborX_DetailsSortUtils.hpp [hunk 6]
```diff
@@ -114,6 +114,67 @@ Kokkos::View<size_t *, Kokkos::Device<Kokkos::Cuda, MemorySpace>> sortObjects(
   return permute;
 }
 #endif
+
+// Helper functions and structs for applyPermutations
+namespace PermuteHelper
+{
+template <class DstViewType, class SrcViewType, int Rank = DstViewType::Rank>
+struct CopyOp;
```
Comment (Rombur) [inline src/details/ArborX_DetailsSortUtils.hpp]: I don't think you need to specialization. I think something like this should work:
```c++
KOKKOS_INLINE_FUNCTION
static void copy(DstViewType const &dst, size_t i_dst, SrcViewType const &src,
                 size_t i_src)
{
  for (unsigned int j = 0; j < dst.extent(1); ++j)
    for (unsigned int k = 0; k < dst.extent(2); ++k)
      for (unsigned int m = 0; m < dst.extent(3); ++m)     
         dst.access(i_dst, j, k, m) = src.access(i_src, j, k, m);
}
```
Note that I am not advocating this solution. I just wanted to point out that this was possible.

**Pull request merged**